src/main/java/htsjdk/samtools/BAMShardIndexer.java

```diff
@@ -0,0 +1,190 @@
+package htsjdk.samtools;
+
+import htsjdk.samtools.util.Log;
+
+import java.io.File;
+import java.io.OutputStream;
+
+public class BAMShardIndexer {
+  // output written as binary, or (for debugging) as text
+  private final BinaryBAMShardIndexWriter outputWriter;
+
+  // content is built up from the input bam file using this
```

Comment:
No problem, but it would be nice to reference what was taken from other places.  Not everything I assume is from HTSJDK in my comments.  The `pom.xml` for instance would probably also need updating regarding the latest updates to packages, which might make the code run even faster.  

What you guys are doing is really important work to get high-throughput sequencing analysis to the next level, and proper documentation and consistent code will make reading and using the code more seamless and ubiquitous.  This way things would not need to be addressed later on in multiple places, which can become a headache.  I can take look again after the fixes.  I'm just helping you out man, as I've gone through things like this before :)